### PR TITLE
Add RDN.

### PR DESCRIPTION
**Pull request by Yshuo-Li**
Residual Dense Network for Image Super-Resolution.

### CODE REVIEW

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 1]
```diff
@@ -0,0 +1,187 @@
...
+                 num_blocks=16,
+                 upscale_factor=4,
+                 num_layers=8,
+                 growth_rate=64):
+
+        super(RDN, self).__init__()
+        self.G0 = mid_channels
+        self.G = growth_rate
+        self.D = num_blocks
+        self.C = num_layers
```
Comment (nbei) [inline mmedit/models/backbones/sr_backbones/rdn.py:103]: Such names (`G0`, `G` or `D`) are so ugly and confusing. Please rename it with regular terms.

Comment (nbei) [reply mmedit/models/backbones/sr_backbones/rdn.py:103]: In general, the name of variables should not contain the capital words.

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 2]
```diff
@@ -0,0 +1,187 @@
...
+    """Dense layer
+
+    Args:
+        in_channels (int): Channel number of inputs.
+        out_channels (int): Channel number of outputs.
+
+    """
+
+    def __init__(self, in_channels, out_channels):
+        super(DenseLayer, self).__init__()
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:19]: `super().__init__()`

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        self.conv = nn.Conv2d(
+            in_channels, out_channels, kernel_size=3, padding=3 // 2)
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:21]: So the padding is hardcoded for kernel 3?

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        self.relu = nn.ReLU(inplace=True)
+
+    def forward(self, x):
+        """Forward function.
+
+        Args:
+            x (Tensor): Input tensor with shape (n, c, h, w).
+
+        Returns:
+            Tensor: Forward results, tensor with shape (n, c, h, w).
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:31]: I don't think the shape does not change after concatenating

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+class RDB(nn.Module):
+    """Residual Dense Block of Residual Dense Network
+
+    Args:
+        in_channels (int): Channel number of inputs.
+        out_channels (int): Channel number of outputs.
+    """
+
+    def __init__(self, in_channels, growth_rate, num_layers):
+        super(RDB, self).__init__()
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:45]: `super().__init__()`

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 6]
```diff
@@ -0,0 +1,187 @@
...
+        """
+        return torch.cat([x, self.relu(self.conv(x))], 1)
+
+
+class RDB(nn.Module):
+    """Residual Dense Block of Residual Dense Network
+
+    Args:
+        in_channels (int): Channel number of inputs.
+        out_channels (int): Channel number of outputs.
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:41]: docstring does not match `__init__`

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    """
+
+    def __init__(self, in_channels, growth_rate, num_layers):
+        super(RDB, self).__init__()
+        self.layers = nn.Sequential(*[
+            DenseLayer(in_channels + growth_rate * i, growth_rate)
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:47]: is growth_rate an int or a float ratio?

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 8]
```diff
@@ -0,0 +1,187 @@
...
+
+
+@BACKBONES.register_module()
+class RDN(nn.Module):
+    """RDN model for single image super-resolution.
+
+    Paper: Residual Dense Network for Image Super-Resolution
+    Adapted from:
+        https://github.com/yulunzhang/RDN.git
+        https://github.com/yjn870/RDN-pytorch
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:74]: Add licence info

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 9]
```diff
@@ -0,0 +1,187 @@
...
+        in_channels (int): Channel number of inputs.
+        out_channels (int): Channel number of outputs.
+        mid_channels (int): Channel number of intermediate features.
+            Default: 64.
+        num_blocks (int): Block number in the trunk network. Default: 16.
+        upscale_factor (int): Upsampling factor. Support 2^n and 3.
+            Default: 4.
+        num_layer (int): Layer number in the Residual Dense Block.
+            Default: 8.
+        growth_rate(int): Channels growth in each layer of RDB.
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:86]: `rate`, a terminology that is often related to percentage, is not good to describe this number.

How about just `growth`, or `channel_growth` or `growth_num`, etc.

File: mmedit/models/backbones/sr_backbones/rdn.py (hunk 1 above)
Comment (innerlee) [reply mmedit/models/backbones/sr_backbones/rdn.py:103]: Yeah, just use `self.mid_channels = mid_channels`

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 10]
```diff
@@ -0,0 +1,187 @@
...
+        else:
+            self.upscale = nn.Sequential(
+                nn.Conv2d(
+                    self.G0,
+                    self.G0 * (upscale_factor**2),
+                    kernel_size=3,
+                    padding=3 // 2), nn.PixelShuffle(upscale_factor))
+
+        self.output = nn.Conv2d(
+            self.G0, out_channels, kernel_size=3, padding=3 // 2)
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:144]: another hardcoded padding

Comment (innerlee) [reply mmedit/models/backbones/sr_backbones/rdn.py:144]: nvmd, the kernel is hardcoded, too

File: mmedit/models/backbones/sr_backbones/rdn.py (hunk 3 above)
Comment (innerlee) [reply mmedit/models/backbones/sr_backbones/rdn.py:21]: nvmd

**Comment by innerlee** [comment]
Mostly some comments on naming. Also need to rebase master :)

File: mmedit/models/backbones/sr_backbones/rdn.py (hunk 1 above)
Comment (Yshuo-Li) [reply mmedit/models/backbones/sr_backbones/rdn.py:103]: `G`, `D`, `C` e.t. are variables in the paper, I'll change them in code.

File: mmedit/models/backbones/sr_backbones/rdn.py (hunk 7 above)
Comment (Yshuo-Li) [reply mmedit/models/backbones/sr_backbones/rdn.py:47]: It is `int`, which is the channel number of dense layer output.

File: mmedit/models/backbones/sr_backbones/rdn.py (hunk 8 above)
Comment (Yshuo-Li) [reply mmedit/models/backbones/sr_backbones/rdn.py:74]: These GitHub projects don't have license files.

File: mmedit/models/backbones/sr_backbones/rdn.py (hunk 9 above)
Comment (Yshuo-Li) [reply mmedit/models/backbones/sr_backbones/rdn.py:86]: picked `channel_growth`.

File: mmedit/models/backbones/sr_backbones/rdn.py (hunk 8 above)
Comment (innerlee) [reply mmedit/models/backbones/sr_backbones/rdn.py:77]: Okay, need to contact them

**Comment by innerlee** [comment]
Conflict with master now

File: mmedit/models/backbones/sr_backbones/rdn.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+        return x + self.lff(self.layers(x))  # local residual learning
+
+
+@BACKBONES.register_module()
+class RDN(nn.Module):
+    """RDN model for single image super-resolution.
+
+    Paper: Residual Dense Network for Image Super-Resolution
+
+    Adapted from 'git@github.com:yjn870/RDN-pytorch.git'
```
Comment (innerlee) [inline mmedit/models/backbones/sr_backbones/rdn.py:76]: use https://github.com/...

**Review by innerlee — APPROVED** [review]

**Comment by innerlee** [comment]
Linting fails
